packages/backend/src/accounting/psql/balance.ts

```diff
@@ -16,38 +14,45 @@ export async function getAccountBalances(
   account: LedgerAccount,
   trx?: TransactionOrKnex
 ): Promise<AccountBalance> {
-  const { credits, debits } = await getAccountTransfers(
-    deps,
-    account.id,
-    undefined, // No limit for balances
-    trx
-  )
+  try {
+    const queryResult = await (trx ?? deps.knex).raw(
+      `
+    SELECT
+      COALESCE(SUM("amount") FILTER(WHERE "creditAccountId" = :accountId AND "state" = 'POSTED'), 0) AS "creditsPosted",
+      COALESCE(SUM("amount") FILTER(WHERE "creditAccountId" = :accountId AND "state" = 'PENDING'), 0) AS "creditsPending",
+      COALESCE(SUM("amount") FILTER(WHERE "debitAccountId" = :accountId AND "state" = 'POSTED'), 0) AS "debitsPosted",
+      COALESCE(SUM("amount") FILTER(WHERE "debitAccountId" = :accountId AND "state" = 'PENDING'), 0) AS "debitsPending"
+    FROM "ledgerTransfers"
+    WHERE ("creditAccountId" = :accountId OR "debitAccountId" = :accountId)
+      AND ("state" = 'POSTED' OR ("state" = 'PENDING' AND "expiresAt" > NOW()));
+    `,
+      { accountId: account.id }
+    )
 
-  let creditsPosted = 0n
-  let creditsPending = 0n
-  let debitsPosted = 0n
-  let debitsPending = 0n
-
-  for (const credit of credits) {
-    if (credit.state === LedgerTransferState.POSTED) {
-      creditsPosted += credit.amount
-    } else if (credit.state === LedgerTransferState.PENDING) {
-      creditsPending += credit.amount
+    if (queryResult?.rows < 1) {
+      throw new Error('No results when fetching balance for account')
```

Comment:
I was always wondered how do we decide if we should throw error, or just return empty response (in case of a balance, 0)?

Reply:
I think your question still stands but I played around with it because I wasnt sure how it would work and wanted to add some context. When there are no records with the associated id it still aggregates a result, its just 0 for all the balances. So I think like Max mentioned in his other comment, this could probably never happen. But yeah I think it still begs the question, do we need to error or can we just return 0 balances.

Reply:
This check is if results of the query weren't able to be returned at all - so missing information unexpectedly, a result of a psql issue. So in that case we should break everything IMO